sdks/python/apache_beam/io/filesystem_test.py

```diff
@@ -0,0 +1,233 @@
+import os
+import unittest
+import tempfile
+import bz2
+import gzip
+from random import randint
+from StringIO import StringIO
+
+from parameterized import parameterized
+
+from apache_beam.io.filesystem import CompressedFile, CompressionTypes
+
+
+class _TestCaseWithTempDirCleanUp(unittest.TestCase):
+  """Base class for TestCases that deals with TempDir clean-up.
+
+  Inherited test cases will call self._new_tempdir() to start a temporary dir
+  which will be deleted at the end of the tests (when tearDown() is called).
+  """
+
+  def setUp(self):
+    self._tempdirs = []
+
+  def tearDown(self):
+    for path in self._tempdirs:
+      if os.path.exists(path):
+        shutil.rmtree(path)
+    self._tempdirs = []
+
+  def _new_tempdir(self):
+    result = tempfile.mkdtemp()
+    self._tempdirs.append(result)
+    return result
+
+  def _create_temp_file(self, name='', suffix=''):
+    if not name:
+      name = tempfile.template
+    file_name = tempfile.NamedTemporaryFile(
+        delete=False, prefix=name,
+        dir=self._new_tempdir(), suffix=suffix).name
+    return file_name
+
+
+class TestCompressedFile(_TestCaseWithTempDirCleanUp):
+  content = """- the BEAM -
+How things really are we would like to know.
+Does
+     Time
+          flow, is it elastic, or is it
+atomized in instants hammered around the
+    clock's face? ...
+- May Swenson"""
+
+  # Keep the read block size small so that we exercise the seek functionality
+  # in compressed file and not just in the internal buffer
+  read_block_size = 4
+
+  def _create_compressed_file(self, compression_type, content,
+                              name='', suffix=''):
+    file_name = self._create_temp_file(name, suffix)
+
+    if compression_type == CompressionTypes.BZIP2:
+      compress_factory = bz2.BZ2File
+    elif compression_type == CompressionTypes.GZIP:
+      compress_factory = gzip.open
+    else:
+      assert False, "Invalid compression type: %s" % compression_type
+
+    with compress_factory(file_name, 'w') as f:
+      f.write(content)
+
+    return file_name
+
+  def test_seekable_enabled_on_read(self):
+    readable = CompressedFile(open(self._create_temp_file(), 'r'))
+    self.assertTrue(readable.seekable)
+
+  def test_seekable_disabled_on_write(self):
+    writeable = CompressedFile(open(self._create_temp_file(), 'w'))
+    self.assertFalse(writeable.seekable)
+
+  def test_seekable_disabled_on_append(self):
+    writeable = CompressedFile(open(self._create_temp_file(), 'a'))
+    self.assertFalse(writeable.seekable)
+
+  @parameterized.expand([CompressionTypes.BZIP2, CompressionTypes.GZIP])
+  def test_seek_set(self, compression_type):
+    file_name = self._create_compressed_file(compression_type, self.content)
+
+    compressed_fd = CompressedFile(open(file_name, 'r'), compression_type,
+                                   read_size=self.read_block_size)
+    reference_fd = StringIO(self.content)
+
+    random_position = randint(0, len(self.content) - 1)
```

Comment:
Let's run this for several fixed positions within the range (0, size - 1) instead of a random position to make the test consistent.